frontendGraph/src/main/java/com/virtuslab/gitmachete/frontend/graph/print/SimpleGraphCellPainter.java

```diff
@@ -0,0 +1,189 @@
+package com.virtuslab.gitmachete.frontend.graph.print;
+
+import java.awt.BasicStroke;
+import java.awt.Color;
+import java.awt.Graphics2D;
+import java.awt.RenderingHints;
+import java.awt.geom.Ellipse2D;
+import java.util.Collection;
+
+import com.intellij.vcs.log.graph.EdgePrintElement;
+import com.intellij.vcs.log.graph.NodePrintElement;
+import com.intellij.vcs.log.graph.PrintElement;
+import com.intellij.vcs.log.paint.ColorGenerator;
+import com.intellij.vcs.log.paint.GraphCellPainter;
+import com.intellij.vcs.log.paint.PaintParameters;
+import com.intellij.vcs.log.paint.PositionUtil;
+import lombok.RequiredArgsConstructor;
+import org.checkerframework.checker.guieffect.qual.UIEffect;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+@RequiredArgsConstructor
+public class SimpleGraphCellPainter implements GraphCellPainter {
+
+  private final ColorGenerator myColorGenerator;
+
+  @UIEffect
+  protected int getRowHeight() {
+    return PaintParameters.ROW_HEIGHT;
+  }
+
+  @UIEffect
+  private BasicStroke getOrdinaryStroke() {
+    return new BasicStroke(PaintParameters.getLineThickness(getRowHeight()), BasicStroke.CAP_ROUND,
+        BasicStroke.JOIN_BEVEL);
+  }
+
+  @UIEffect
+  private void paintUpLine(Graphics2D g2, Color color, int from, int to) {
+    // paint vertical lines normal size
+    // paint non-vertical lines twice the size to make them dock with each other well
+    int nodeWidth = PaintParameters.getNodeWidth(getRowHeight());
+    if (from == to) {
+      int x = nodeWidth * from + nodeWidth / 2;
+      int y1 = getRowHeight() / 2 - 1;
+      int y2 = 0;
+      paintLine(g2, color, x, y1, x, y2);
+    } else {
+      int x1 = nodeWidth * from + nodeWidth / 2;
+      int y1 = getRowHeight() / 2;
+      int x2 = nodeWidth * to + nodeWidth / 2;
+      int y2 = -getRowHeight() / 2;
+      paintLine(g2, color, x1, y1, x2, y2);
+    }
+  }
+
+  @UIEffect
+  private void paintDownLine(Graphics2D g2, Color color, int from, int to) {
+    int nodeWidth = PaintParameters.getNodeWidth(getRowHeight());
+    if (from == to) {
+      int y2 = getRowHeight();
+      int y1 = getRowHeight() / 2;
+      int x = nodeWidth * from + nodeWidth / 2;
+      paintLine(g2, color, x, y1, x, y2);
+    } else {
+      int x1 = nodeWidth * from + nodeWidth / 2;
+      int y1 = getRowHeight() / 2;
+      int x2 = nodeWidth * to + nodeWidth / 2;
+      int y2 = getRowHeight() + getRowHeight() / 2;
+      paintLine(g2, color, x1, y1, x2, y2);
+    }
+  }
+
+  @UIEffect
+  private void paintRightLine(Graphics2D g2, Color color, int from, int to) {
```

Comment:
Out of curiosity... why no `paintLeftLine`?...

Reply:
Cause (unlike the up and down lines) we can draw it within a single cell. For a given `row`, we have a cell, for a given `cell` we have graph elements (nodes and edges) aligned to their positions. The right line is a line just between position `x` (graph element position) and `x+1`. There is no need to split it into left and right line (or providing `x` to `x-1` line which might cause bugs fo `x=0`).